File: src/c++/msgbus/metrics.cpp
#include "metrics.hpp"

namespace otsim {
namespace msgbus {

void MetricsPusher::Start(std::shared_ptr<Pusher> pusher, const std::string& name) {
  thread = std::thread(&MetricsPusher::run, this, pusher, name);
}

void MetricsPusher::Stop() {
  running.store(false);

  if (thread.joinable()) {
    thread.join();
  }
}

void MetricsPusher::NewMetric(const std::string& kind, const std::string& name, const std::string& desc) {
  Metric metric = {
    .kind = kind,
    .name = name,
    .desc = desc,
  };

  metrics[name] = metric;
}

void MetricsPusher::IncrMetric(const std::string &name) {
  try {
    auto lock = std::unique_lock<std::mutex>(metricsMu);

    auto metric = metrics.at(name);
    metric.value += 1.0;

    metrics[name] = metric;
  } catch(const std::out_of_range&) {}
}

void MetricsPusher::IncrMetricBy(const std::string &name, int val) {
  try {
    auto lock = std::unique_lock<std::mutex>(metricsMu);

    auto metric = metrics.at(name);
    metric.value += val;

    metrics[name] = metric;
  } catch(const std::out_of_range&) {}
}

void MetricsPusher::SetMetric(const std::string &name, double val) {
  try {
    auto lock = std::unique_lock<std::mutex>(metricsMu);

    auto metric = metrics.at(name);
    metric.value = val;

    metrics[name] = metric;
  } catch(const std::out_of_range&) {}
}

void MetricsPusher::run(std::shared_ptr<Pusher> pusher, const std::string& name) {
  auto prefix = name + "_";

  running.store(true);

  while (running) {
    std::vector<Metric> updates;

    {
      auto lock = std::unique_lock<std::mutex>(metricsMu);

      for (auto [name, metric] : metrics) {
        auto copy = metric;

        auto pos = std::mismatch(prefix.begin(), prefix.end(), copy.name.begin());
        if (pos.first != name.end()) {
          copy.name = prefix + copy.name;
        }

        updates.push_back(copy);
      }
    }

    if (updates.size() > 0) {
      auto env = NewEnvelope(name, Metrics{.metrics = updates});
      pusher->Push("HEALTH", env);
    }

    std::this_thread::sleep_for(std::chrono::seconds(5));
  }
}

} // namespace msgbus
} // namespace otsim

File: src/python/otsim/ieee_2030_5/msgbus/metrics.py
from __future__ import annotations

import json, threading, time, typing

import otsim.msgbus.envelope as envelope

from otsim.msgbus.envelope import Metric, MetricKind
from otsim.msgbus.pusher   import Pusher

class MetricsPusher:
  def __init__(self: MetricsPusher):
    self.running = False
    self.metrics: typing.Dict[str, Metric] = {}

  def start(self: MetricsPusher, pusher: Pusher, name: str) -> None:
    self.running = True

    self.thread = threading.Thread(target=self.__run, args=(pusher, name,))
    self.thread.start()

  def stop(self: MetricsPusher) -> None:
    self.running = False
    self.thread.join()

  def new_metric(self: MetricsPusher, kind: MetricKind, name: str, desc: str) -> None:
    self.metrics[name] = {'kind': kind.value, 'name': name, 'desc': desc, 'value': 0.0}

  def incr_metric(self: MetricsPusher, name: str) -> None:
    if name in self.metrics:
      metric = self.metrics[name]
      metric['value'] += 1.0
      self.metrics[name] = metric

  def incr_metric_by(self: MetricsPusher, name: str, val: int) -> None:
    if name in self.metrics:
      metric = self.metrics[name]
      metric['value'] += float(val)
      self.metrics[name] = metric

  def set_metric(self: MetricsPusher, name: str, val: float) -> None:
    if name in self.metrics:
      metric = self.metrics[name]
      metric['value'] = val
      self.metrics[name] = metric

  def __run(self: MetricsPusher, pusher: Pusher, name: str) -> None:
    prefix = name + "_"

    while self.running:
      updates: typing.List[Metric] = []

      for metric in self.metrics.values():
        copy = metric

        if not copy['name'].startswith(prefix):
          copy['name'] = prefix + copy['name']

        updates.append(copy)

      if len(updates) > 0:
        env = envelope.new_metric_envelope(name, {'metrics': updates})
        pusher.push('HEALTH', env)

      time.sleep(5)